core/src/main/java/org/infinispan/util/concurrent/BlockingManagerImpl.java

```diff
@@ -207,10 +207,23 @@ public void execute(Runnable command) {
       return continueOnNonBlockingThread(stage.thenRunAsync(runnable, blockingExecutor), traceId);
    }
 
+   @Override
+   public <I, O> CompletionStage<O> thenComposeBlocking(CompletionStage<? extends I> stage,
+         Function<? super I, ? extends CompletionStage<O>> function, Object traceId) {
+      if (CompletionStages.isCompletedSuccessfully(stage) && isCurrentThreadBlocking()) {
+         if (log.isTraceEnabled()) {
+            log.tracef("Invoked thenComposeBlocking on a blocking thread, joining %s in same blocking thread", traceId);
+         }
+         I value = CompletionStages.join(stage);
+         return function.apply(value);
```

Comment:
`try { ... }  catch { ... }` missing IMO

Reply:
Ah yes it is now that I am not using thenCompose anymore :(